src/Cmts.ml

```diff
@@ -516,24 +524,56 @@ let fmt_cmt t cmt =
 let fmt_cmts t ?pro ?epi ?(eol= Fmt.fmt "@\n") ?(adj= eol) tbl loc =
   let open Fmt in
   let find = if !remove then Hashtbl.find_and_remove else Hashtbl.find in
-  let cmts = Option.value (find tbl loc) ~default:[] in
-  let last_cmt = List.last cmts in
-  let eol_cmt =
-    Option.value ~default:false
-      (last_cmt >>| fun (_, loc) -> Source.ends_line t.source loc)
-  in
-  let adj_cmt =
-    eol_cmt
-    && Option.value ~default:false
-         ( last_cmt
-         >>| fun (_, {Location.loc_end= {pos_lnum}}) ->
-         pos_lnum + 1 = loc.Location.loc_start.pos_lnum )
-  in
-  fmt_if_k
-    (not (List.is_empty cmts))
-    ( Option.call ~f:pro
-    $ vbox 0 (list cmts "@ " (fmt_cmt t))
-    $ fmt_or_k eol_cmt (fmt_or_k adj_cmt adj eol) (Option.call ~f:epi) )
+  match find tbl loc with
+  | None -> fmt ""
+  | Some cmts ->
+      let line_dist a b =
+        b.Location.loc_start.pos_lnum - a.Location.loc_end.pos_lnum
+      in
+      let groups =
+        List.group cmts ~break:(fun (_, a) (_, b) ->
+            not
+              ( Location.is_single_line a && Location.is_single_line b
+              && line_dist a b = 1
+              && Location.compare_start_col a b = 0
+              && Location.compare_end_col a b = 0 ) )
+      in
+      let last_cmt = List.last cmts in
+      let eol_cmt =
+        Option.value ~default:false
+          (last_cmt >>| fun (_, loc) -> Source.ends_line t.source loc)
+      in
+      let adj_cmt =
+        eol_cmt
+        && Option.value ~default:false
+             ( last_cmt
+             >>| fun (_, {Location.loc_end= {pos_lnum}}) ->
+             pos_lnum + 1 = loc.Location.loc_start.pos_lnum )
+      in
+      let maybe_newline ~next (_, cur_last_loc) =
+        match next with
+        | Some ((_, next_loc) :: _) ->
+            fmt_if (line_dist cur_last_loc next_loc > 1) "@;<1000 0>"
+        | _ -> fmt ""
+      in
+      list_pn groups (fun ?prev group ?next ->
+          fmt_or_k (Option.is_none prev)
+            (Option.call ~f:pro $ open_vbox 0)
+            (fmt "@ ")
+          $ ( match group with
+            | [] -> impossible "previous match"
+            | [cmt] -> fmt_cmt t cmt $ maybe_newline ~next cmt
+            | group ->
+                fmt_if_k (Option.is_none prev)
+                  (Fmt.break_unless_newline 0 0)
```

Comment:
Do you have examples where the 2 lines above are needed? With `--sparse` they lead to an extra open line before the comments in the following example:
```
let f x =
  match x with
  | `A ->
      ()
  | `B ->
      (* Proin ipsum nunc, finibus et finibus, semper et mi. Aenean *)
      (* pretium fermentum tellus, a faucibus sagittis et. Cras non *)
      ()
```
The `Format` operations that are conditional on newlines are pretty fragile to use, since leading whitespace for indentation counts as being not starting a new line.

This is the last little issue I have seen in my testing.

Reply:
sorry, I've pushed -force and removed your commit.

Reply:
These two lines are not need and are now removed.
I've added back your example.